include/oneapi/dpl/pstl/hetero/dpcpp/execution_sycl_defs.h

```diff
@@ -34,29 +36,75 @@ inline namespace __dpl
 
 struct DefaultKernelName;
 
+////////////////////////////////////////////////////////////////////////////////
+// sycl_queue_container - container with optionally created sycl::queue instance
+template <class TSYCLQueueFactory>
+class sycl_queue_container
+{
+  public:
+    template <typename... Args>
+    sycl::queue
+    get_queue(Args&&... args)
+    {
+        ::std::call_once(__is_created, [&]() {
+            TSYCLQueueFactory factory;
+            __queue.emplace(factory(::std::forward<Args>(args)...));
+        });
+
+        assert(__queue.has_value());
+        return __queue.value();
+    }
+
+  private:
+    ::std::once_flag __is_created;
+    ::std::optional<sycl::queue> __queue;
+};
+template <typename TFactory>
+using sycl_queue_container_ptr = ::std::shared_ptr<sycl_queue_container<TFactory>>;
+
+////////////////////////////////////////////////////////////////////////////////
+// sycl_queue_factory_device - default sycl::queue factory for device policy
+struct sycl_queue_factory_device
+{
+    template <typename... Args>
+    sycl::queue
+    operator()(Args&&... args)
+    {
+        return sycl::queue(::std::forward<Args>(args)...);
+    }
+};
+
 //We can create device_policy object:
 // 1. from sycl::queue
 // 2. from sycl::device_selector (implicitly through sycl::queue)
 // 3. from sycl::device
 // 4. from other device_policy encapsulating the same queue type
-template <typename KernelName = DefaultKernelName>
+template <typename KernelName = DefaultKernelName, class TSyclQueueFactory = sycl_queue_factory_device>
```

Comment:
As far as I can see in [oneDPL specification](https://spec.oneapi.io/versions/latest/elements/oneDPL/source/parallel_api.html#dpc-execution-policy) - `device_policy` class is a public API and hence its template parameters are also documented. So do we think that it is necessary to have `TSyclQueueFactory` template parameter as part of public API?

Reply:
In this valiant - probably yes.